microdnf/utils.cpp

```diff
@@ -48,31 +52,47 @@ std::filesystem::path get_user_home_dir() {
 }
 
 std::filesystem::path get_user_cache_dir() {
-    if (char * dir = std::getenv("XDG_CACHE_HOME")) {
-        return std::filesystem::path(dir);
+    const char * dir = std::getenv("XDG_CACHE_HOME");
+    if (dir && *dir != '\0') {
+        auto ret = std::filesystem::path(dir);
+        if (ret.is_absolute()) {
+            return ret;
+        }
     }
     return std::filesystem::path(get_user_home_dir()) / ".cache";
 }
 
 std::filesystem::path get_user_config_dir() {
-    if (char * dir = std::getenv("XDG_CONFIG_HOME")) {
-        return std::filesystem::path(dir);
+    const char * dir = std::getenv("XDG_CONFIG_HOME");
+    if (dir && *dir != '\0') {
+        auto ret = std::filesystem::path(dir);
+        if (ret.is_absolute()) {
+            return ret;
+        }
     }
     return std::filesystem::path(get_user_home_dir()) / ".config";
 }
 
 std::filesystem::path get_user_data_dir() {
-    if (char * dir = std::getenv("XDG_DATA_HOME")) {
-        return std::filesystem::path(dir);
+    const char * dir = std::getenv("XDG_DATA_HOME");
+    if (dir && *dir != '\0') {
+        auto ret = std::filesystem::path(dir);
+        if (ret.is_absolute()) {
+            return ret;
+        }
     }
     return std::filesystem::path(get_user_home_dir()) / ".local" / "share";
 }
 
 std::filesystem::path get_user_runtime_dir() {
-    if (char * dir = std::getenv("XDG_RUNTIME_HOME")) {
-        return std::filesystem::path(dir);
+    const char * dir = std::getenv("XDG_RUNTIME_DIR");
+    if (dir && *dir != '\0') {
+        auto ret = std::filesystem::path(dir);
+        if (ret.is_absolute()) {
+            return ret;
+        }
     }
-    return get_user_cache_dir();
+    throw std::runtime_error("get_user_runtime_dir(): Can't determine the user's runtime directory");
```

Comment:
Given these are internal to libdnf (after moving them to libdnf/utils), I'm not heavily opposed to having them, but do we have a use for them? The `get_user_runtime_dir()` in particular, which is set by pam for the session and has no default. I'm not sure what we would use it for, it seems dnf shouldn't depend on the session working (though I'm not even sure if you can login without that nowadays?), in case the user sessions break we still want dnf to be functional to fix the issue, right?

Right now only the cache dir is used, I'd be inclined to only add those that we use or know we will need in the future. If the others are needed later, they can be added?